eyeball-im/tests/it/batch.rs

```diff
@@ -0,0 +1,42 @@
+use imbl::vector;
+use stream_assert::{assert_next_eq, assert_pending};
+
+use eyeball_im::{ObservableVector, VectorDiff};
+
+#[test]
+fn lagging_batch_stream() {
+    let mut ob = ObservableVector::new();
+    let mut st = ob.subscribe().into_batched_stream();
+
+    ob.push_back(0);
+    ob.append(vector![1, 2]);
+    ob.push_back(3);
+
+    assert_next_eq!(
+        st,
+        vec![
+            VectorDiff::PushBack { value: 0 },
+            VectorDiff::Append { values: vector![1, 2] },
+            VectorDiff::PushBack { value: 3 },
+        ]
+    );
+}
+
+#[test]
+fn transaction() {
+    let mut ob = ObservableVector::new();
+    let mut st = ob.subscribe().into_batched_stream();
+    let mut txn = ob.transaction();
+
+    txn.push_back(0);
+    assert_pending!(st);
+
+    txn.push_front(-1);
+    assert_pending!(st);
+
+    drop(txn);
+    assert_next_eq!(
+        st,
+        vec![VectorDiff::PushBack { value: 0 }, VectorDiff::PushFront { value: -1 }]
+    );
+}
```

Comment:
You may also want to test that when a transaction is “cancelled”, nothing is applied. Is it even possible? I don't think so regarding the current state of the patch.

Idea:

Implement `Transaction::apply` to _apply_ the diffs in it, and implement `Drop for Transaction` to just abort/“cancel” the transaction. Thoughts?

Taking a step back, it indeed looks weird that dropping a transaction applies its diffs. A user may expect that it cancels/aborts the transaction instead.

Reply:
> You may also want to test that when a transaction is “cancelled”, nothing is applied. Is it even possible? I don't think so regarding the current state of the patch.

Yeah, transaction rollback / cancellation is not supported, as documented. It shouldn't be that hard to add but then the question arises of what should be the default behavior when the transaction is dropped and so on. And we don't need it for the Matrix Rust SDK, so I decided to punt on it for now and document that.